packages/@react-aria/overlays/src/calculatePosition.ts

```diff
@@ -313,6 +313,27 @@ export function calculatePositionInternal(
     }
   }
 
+  let boundaryLimits = {
+    top: boundaryDimensions.top + boundaryDimensions.scroll.top,
+    bottom: boundaryDimensions.top + boundaryDimensions.scroll.top + boundaryDimensions.height,
+    left: boundaryDimensions.left + boundaryDimensions.scroll.left,
+    right: boundaryDimensions.left + boundaryDimensions.scroll.left + boundaryDimensions.width
+  };
+
+  // for popover triggers close to the edge decrease the padding
+  if (placementInfo.placement === 'left' || placementInfo.placement === 'right') {
+    let centerOfButtonVertical = childOffset.top + childOffset.height / 2;
+    // Really it should be center of button center + padding + half of arrow width > boundary but we don't have access to arrow dimensions
+    if (centerOfButtonVertical + padding * 2 > boundaryLimits.bottom || centerOfButtonVertical - padding * 2 < boundaryLimits.top) {
+      padding = 6; // setting the padding to 6px as defined by Spectrum for this case
```

Comment:
Sorry to interrupt, but is it true that we override user-defined padding value for popover (not only arrow) in some cases? For example, if I set padding value to 1 and it perfectly fits my needs and trigger gets close to boundary, padding will be overridden to 6?

Reply:
I did a quick experiment and found that setting the `containerPadding` to 2px on edges did not trigger this. If my testing missed something, please let me know. The immediate thought of a fix would be only doing this block of logic if `padding == DEFAULT_MODAL_PADDING`.